### PR TITLE
Feat/v1.12/cilium strict mode

### PR DESCRIPTION
**Pull request by 3u13r**
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

### CODE REVIEW

File: bpf/lib/encap.h [hunk 1]
```diff
@@ -225,8 +228,19 @@ encap_and_redirect_lxc(struct __ctx_buff *ctx, __u32 tunnel_endpoint,
 	}
 
 	tunnel = map_lookup_elem(&TUNNEL_MAP, key);
-	if (!tunnel)
+	if (!tunnel){
 		return DROP_NO_TUNNEL_ENDPOINT;
+	}
+
+#ifdef TUNNEL_MODE
```
Comment (benschlueter) [inline bpf/lib/encap.h]: group defines

File: bpf/bpf_host.c [hunk 2]
```diff
@@ -1103,6 +1108,16 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 		}
 	}
 
+#ifndef TUNNEL_MODE
```
Comment (benschlueter) [inline bpf/bpf_host.c]: group defines

File: bpf/lib/strict.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+strict_allow(struct __ctx_buff *ctx)
+{
+    void *data, *data_end;
+    __u16 proto = 0;
+    __u64 length = 0, offset = 0;
+    bool __maybe_unused in_strict_cidr = false;
+    struct tcphdr *tcph;
+#ifdef ENABLE_IPV6
+    struct ipv6hdr *ip6;
+#endif
```
Comment (benschlueter) [inline bpf/lib/strict.h]: unused
```suggestion
```

File: bpf/lib/strict.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                offset = sizeof(struct ethhdr) + sizeof(struct iphdr);
+                if ((data + offset + sizeof(struct tcphdr)) > data_end)
+                    return !in_strict_cidr;
+                tcph = (struct tcphdr *)(data + offset);
+                length -= (tcph->doff << 2);
+
+                printk("TCP TCP TCP source %pI4, dest %pI4\n", &ip4->saddr, &ip4->daddr);
+                printk("tcp length %llx, source %d, dest %d\n", length, bpf_htons(tcph->source), bpf_htons(tcph->dest));
+            }
+
```
Comment (benschlueter) [inline bpf/lib/strict.h]: unused
```suggestion
```

File: bpf/lib/strict.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            return !in_strict_cidr;
+        case IPPROTO_UDP:
+            if (in_strict_cidr)
+            {
+                printk("UDP UDP UDP source %pI4, dest %pI4\n", &ip4->saddr, &ip4->daddr);
+            }
```
Comment (benschlueter) [inline bpf/lib/strict.h]: unused
```suggestion
```

File: bpf/lib/strict.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+#ifndef __WIREGUARD_H_
+#define __WIREGUARD_H_
```
Comment (benschlueter) [inline bpf/lib/strict.h]:
```suggestion
#ifndef __STRICT_H_
#define __STRICT_H_
```

File: install/kubernetes/cilium/values.yaml [hunk 7]
```diff
@@ -547,6 +547,8 @@ encryption:
   # This option is only effective when encryption.type is set to ipsec.
   interface: ""
 
+strictModeCIDR: ""
```
Comment (Kakashiiiiy) [inline install/kubernetes/cilium/values.yaml:550]:
```suggestion
strictModeCIDR: []
```

File: pkg/datapath/linux/config/utils.go [hunk 8]
```diff
@@ -60,3 +61,11 @@ func defineMAC(name string, addr []byte) string {
...
 		name, common.GoArray2C(addr), name, name)
 }
+
+func ipv4ToInt(ip net.IP) uint32 {
+	fourByteIP := ip.To4()
+	if fourByteIP == nil {
+		return 0
+	}
+	return binary.LittleEndian.Uint32(fourByteIP)
+}
```
Comment (benschlueter) [inline pkg/datapath/linux/config/utils.go]: unused
```suggestion
```

File: pkg/datapath/linux/config/utils_test.go [hunk 9]
```diff
@@ -76,3 +77,7 @@ func (s *ConfigSuite) TestdefineMAC(c *C) {
 		c.Assert(defineMAC("foo", test.input), Equals, test.output)
 	}
 }
+
+func (s *ConfigSuite) Testipv4ToInt(c *C) {
+	c.Assert(ipv4ToInt(net.IPv4(255, 254, 253, 252)), Equals, 0xfcfdfeff)
+}
```
Comment (benschlueter) [inline pkg/datapath/linux/config/utils_test.go]: unused
```suggestion
```

File: bpf/lib/strict.h [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        break;
+#endif
+    default:
+        return true;
+    }
+}
+
+#endif /* __WIREGUARD_H_ */
```
Comment (benschlueter) [inline bpf/lib/strict.h]: can we delete this `#endif`?

File: pkg/datapath/linux/config/config.go [hunk 11]
```diff
@@ -387,13 +396,11 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 			cDefinesMap["ENABLE_SRC_RANGE_CHECK"] = "1"
 			if option.Config.EnableIPv4 {
 				cDefinesMap["LB4_SRC_RANGE_MAP"] = lbmap.SourceRange4MapName
-				cDefinesMap["LB4_SRC_RANGE_MAP_SIZE"] =
-					fmt.Sprintf("%d", lbmap.SourceRange4Map.MapInfo.MaxEntries)
+				cDefinesMap["LB4_SRC_RANGE_MAP_SIZE"] = fmt.Sprintf("%d", lbmap.SourceRange4Map.MapInfo.MaxEntries)
```
Comment (benschlueter) [inline pkg/datapath/linux/config/config.go]: unnecessary change

File: pkg/datapath/linux/config/config.go [hunk 12]
```diff
@@ -387,13 +396,11 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
...
 			}
 			if option.Config.EnableIPv6 {
 				cDefinesMap["LB6_SRC_RANGE_MAP"] = lbmap.SourceRange6MapName
-				cDefinesMap["LB6_SRC_RANGE_MAP_SIZE"] =
-					fmt.Sprintf("%d", lbmap.SourceRange6Map.MapInfo.MaxEntries)
+				cDefinesMap["LB6_SRC_RANGE_MAP_SIZE"] = fmt.Sprintf("%d", lbmap.SourceRange6Map.MapInfo.MaxEntries)
```
Comment (benschlueter) [inline pkg/datapath/linux/config/config.go]: unnecessary change

File: pkg/datapath/linux/config/config.go [hunk 13]
```diff
@@ -520,8 +527,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		if option.Config.EnableIPv4Masquerade && option.Config.EnableBPFMasquerade {
 			cDefinesMap["ENABLE_MASQUERADE"] = "1"
 			cidr := datapath.RemoteSNATDstAddrExclusionCIDRv4()
-			cDefinesMap["IPV4_SNAT_EXCLUSION_DST_CIDR"] =
-				fmt.Sprintf("%#x", byteorder.NetIPv4ToHost32(cidr.IP))
+			cDefinesMap["IPV4_SNAT_EXCLUSION_DST_CIDR"] = fmt.Sprintf("%#x", byteorder.NetIPv4ToHost32(cidr.IP))
```
Comment (benschlueter) [inline pkg/datapath/linux/config/config.go]: unnecessary change

**Review by benschlueter — CHANGES_REQUESTED** [review]

File: bpf/lib/strict.h (hunk 10 above)
Comment (3u13r) [reply bpf/lib/strict.h]: Nope

**Review by benschlueter — APPROVED** [review]

**Pull request closed without merging**